elkscmd/sys_utils/beep.c

```diff
@@ -69,15 +69,28 @@ void beep_signal(int sig)
 int main(int ac, char **av)
 {
     long duration = 333L;
+    int freq = 1000;
+    int i;
     if(ac >= 2) {
-        duration = atoi(av[1]);
+         for(i = 1; i < ac; i++) {
+              if(av[i][0] = '-') {
+                   switch(av[i][1]) {
+                   case 'f':                    /* Frequency */
+                       freq = atoi(&av[i][2]);
+                       break;
+                   case 'l':                    /* Duration in millisecond */
+                       duration = atoi(&av[i][2]);
```

Comment:
Since `duration` is declared `long`, this should probably use `atol` here. A similar issue could apply with `freq`, which is only valid to 32767. So `beep` and `freq` might want to be declared `beep(unsigned int freq)`. 

In order to keep the code file quite small, one might want to use `atol` in both places.

Reply:
Now I notice that `freq` is used as following in `beep`:
```
d = 2457600 / freq;
```
which means `freq` will have to be converted to long anyways. Thus `freq` should probably be declared long in both beep and the main code, and also have the above rewritten to:
```
d = 2457600L / freq;
```
The reason for this is that the `ia16-elf-gcc` compiler sometimes does the wrong thing with int constants that are > 65535.

Reply:
> The reason for this is that the ia16-elf-gcc compiler sometimes does the wrong thing with int constants that are bigger than 65535.

Hmm, are you sure about this bigger than 65535? I don't remember now exactly, but in main.c I suffixed L to the duration hardcoded value and to the hardcoded value in usleep. And IIRC there was a necessity to it I did not understand at that time (I thought "oh wait 16bit, that's why....?"), both are well below 65535...sorry I cannot verify my theory right now.

Reply:
In C, types that are different sizes (e.g. `char`, `int`, `long`) or are different signedness (e.g. `int` vs `unsigned int`) are automatically promoted to 1) the size of the largest sized element in the expression, and 2) to unsigned if any element is unsigned. Types that are passed to functions are also automatically promoted to the defined argument signedness and size.

For "integer" constants, the question is: what is an integer? For `ia16-elf-gcc`, it is `int`, which is 16-bits. This can mean that the compiler silently truncates a long to an int if there are no other elements in the expression, or if it is not being passed to a pre-defined function. This also occurs with signedness: 65535 requires a long, even though an unsigned int could take that value. So is it signed or unsigned? IIRC, the compiler assumes `int` unless otherwise expressed, which would mean that the value gets truncated to 16 bits. This may or may not be what is desired, as 65535 still fits in 16-bits if the value is interpreted as unsigned. However, if an additional element is type `int` and not `long`, then the arithmetic won't be done in 32 bits.

For these reasons, as well as clarity, in the ELKS kernel, most all literal constants > 32767 are suffixed with `L` so as to ensure proper internal and external computation.  In order to know for sure, one can look at the ASM output using something like:
```
$ ia16-elf-objdump -D -r -Mi8086 beep.o
```

Reply:
Thanks @ghaerr !